### DISCUSSION TITLE
Plugin / rest routing url / PicklingError -> Cache issue

### DISCUSSION THREAD

**Discussion by Bbillyben** [inventree/InvenTree#3948] [Q&A] [answered]
Hi, 

I'm developping my 1st plugin
The plugin class inherit from `PanelMixin`, `SettingsMixin`, `UrlsMixin`,  and `InvenTreePlugin`

UrlsMixin is dedicated here to some rest call to manage bootstrap table datas, the urls are defines as: 
```
    def setup_urls(self):
        router = routers.DefaultRouter()

        router.register(r'track', SMPviews.SMPTrackViewSet, basename='track')
        return router.urls
```

the setting is pretty simple : 

```
    SETTINGS = {
        'MONTH_FOLLOW': {
            'name': 'Number of Month to follow',
            'description': 'time scope in month to visualize from last entry',
            'default': 1,
            'validator': MinValueValidator(1),
        },
    }
```

The panel is added in StockLocation tab to display StokItemTrackin : 
```
    def get_custom_panels(self, view, request):
        if isinstance(view, StockIndex) or isinstance(view, StockLocationDetail):
            
            if isinstance(view, StockLocationDetail):
                loc=view.get_object()
                urlTrac=reverse('plugin:stocktrack:track-location', kwargs={'loc': loc.pk})
            else:
                urlTrac=reverse('plugin:stocktrack:track-list')
                
            context={'dataUrl':urlTrac}
            
            tmpRend = render_to_string(template_name='stocktrackpanel/SMP_panel.html',context=context)

            panels = [
                {
                    # Simple panel without any actual content
                    'title': 'Stock Tracking',
                    'content': tmpRend,
                    'icon' : 'fa-thumbtack', 
                    'javascript_template': 'stocktrackpanel/SMP_panel.js'
                }
            ]
        return panels
```

the viewset inherit from `APIDownloadMixin`, `rest_framework.viewsets.ModelViewSet`: 
with some action to filter, select location and sublocation items, ...

Every thing works fine in StockItem panel, 

But I've got an error in Setting Panel : 
`
_pickle.PicklingError: Can't pickle <function SMPTrackViewSet at 0x7fec3b80d280>: it's not the same object as plugins.StockTrackPanel_plugin.views.SMPTrackViewSet 
`

It's linked to URL routes added and the cache system.

1/ If I route to a direct function rather than the viewset, Setting panel load fine
2/ if I disable cache system by adding in Inventree.setting.py : 
```
CACHES = {
    'default': {
        'BACKEND': 'django.core.cache.backends.dummy.DummyCache',
    }
}
```
Setting panel load fine again.

I've read from gg that it could be due to some import sequence issues, that's why I've split my code in 2 files, but no effect.

=> Have you got any clue to help on that issue?

Versions : 
 Inventree Master last (1104e981a)
in docker dev environnement.

**Comment by matmair** [comment]
@Bbillyben can be connected to a bunch of things. I would suggest trying the following things:
- Turn of debugging - the caching system works a bit differently if caching is enabled; that causes problems sometimes
- use explicit function references or `as_view` pluggers - I honestly have never used viewsets with UrlMixins; once I need custom APIs I use `AppMixin`, there this works
- deepcopy the urlpatterns the router returns - that way they turn static

I am not really sure what you mean by `Setting panel? - could be the time of night here in Europe.
If you have a repo with the code I would be happy to take a look in the coming weeks, might be easier to debug that way.

**Reply by Bbillyben** [reply]
Thanks @matmair,

Yes we were at night here! I would mean Setting page (url `/settings/`).


I've spent a bit of my morning in testing.
1. Turning of debugging do not solve anything.
2. I've fast try to hard code url pattern, but viewset 'static' routing is a pain from my inexperience...
3. If i test with hard coded url, outside viewset, I it seems working => I should rewrite the whole view control.

I will post update as soon as I've move forward on it!

My code (from this night, absolutely no cleaning is available on [github](https://github.com/Bbillyben/StockTrackPanel_plugin))

**Reply by matmair** [reply]
@Bbillyben the gh link 404s

**Reply by Bbillyben** [reply]
Rhaa excuse me, it was in private! I've just switch it

**Comment by Bbillyben** [comment, marked as answer]
I'm back with an ugly solution........... 



## Firstly : Observations


I went to a standard `django.views.View`, for my `SMPTrackViewSet` view class and route by linking .as_view() : 
```
        SMP_URL=[
            path('track/', SMPviews.SMPTrackViewSet.as_view(), name='track-list'),
            path('track/location/<loc>/', SMPviews.SMPTrackViewSet.as_view(), name='track-location'),      
        ]
```
The overall process works fine, but I've got the same error in the 'Setting page' (inventreeURL/settings/)
error trace : [ViewError.txt](https://github.com/inventree/InvenTree/files/10041389/ViewError.txt)

So I test to code as basic as I can and code 'function based view',  to route as : 
```
SMP_URL=[
            path('track/', SMPviews.get_SMP_list, name='track-list'),
            path('track/location/<loc>/', get_SMP_location, name='track-location'),      
        ]
```
I can navigate to /settings/ without any issues

but when i'm trying to access urls, I've got the error :
```
Exception Type: AssertionError at /plugin/stocktrack/track/
Exception Value: Cannot apply DjangoModelPermissions on a view that does not set `.queryset` or have a `.get_queryset()` method.
```
error trace :[FunctionViewError.txt](https://github.com/inventree/InvenTree/files/10041388/FunctionViewError.txt)


## Solution

Set up a standard `django.views.View`
manage your get function.

Add a specific get_smth function to call the get, instanciating the class, eg:
```
    def get_track(request,*args, **kwargs):
        return SMPTrackViewSet().get(request, *args, **kwargs)
```
and route by : 

```
        SMP_URL=[
            path('track/', SMPviews.SMPTrackViewSet.get_track, name='track-list'),
            path('track/location/<loc>/', SMPviews.SMPTrackViewSet.get_track, name='track-location'),      
        ]
```

That's Capillotracted !!!!

Some few test, and I will try to add the plugin in a prod environment!

**Reply by matmair** [reply]
@Bbillyben it might be easier to use AppMixin - interesting finds.

**Reply by Bbillyben** [reply]
@matmair , i've tested added AppMixin in the inherited class of the plugin.

I do not change anything else and I do have the PicklingError

What should I change in urls route definition in your opinion ?

**Reply by matmair** [reply]
@Bbillyben with the AppMixin you can basically just override the whole loading mechanism. I am a bit occupied with release preparations but I will try your repo once I am done with that.

**Reply by Bbillyben** [reply]
Thanks @matmair,
Ok I understand AppMixin will let a plugin define a whole app.
That would be overkill for my purpose here!
Thanks again for your answers. Don't spent much more time on it.
My main concern is [a core js function override](https://github.com/inventree/InvenTree/discussions/3945), as everything run smooth apparently, even in a prod environment!

I'm now trying to inherit from `StockTrackingList` to be more integrated, but I fail on the final `request.is_ajax` that redirect to the response instead of the JsonResponse I'm looking for, wip.

**Reply by matmair** [reply]
@Bbillyben I am currently rewriting the whole UI for #3901 - not sure how long investment in such a function would be helpful.
